persist/writer.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package persist
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/m3db/m3ninx/index/segment"
+	"github.com/m3db/m3ninx/index/segment/fs"
+)
+
+type newFSSegmentWriterFn func() fs.Writer
+
+// NewMutableSegmentFileSetWriter returns a new IndexSegmentFileSetWriter for writing
+// out the provided Mutable Segment.
+func NewMutableSegmentFileSetWriter(
+	seg segment.MutableSegment,
+) (MutableSegmentFileSetWriter, error) {
+	return newMutableSegmentFileSetWriter(fs.NewWriter, seg)
+}
+
+func newMutableSegmentFileSetWriter(
+	fn newFSSegmentWriterFn,
+	seg segment.MutableSegment,
+) (MutableSegmentFileSetWriter, error) {
+	w := &writer{
+		fsWriter: fn(),
```

Comment:
Any reason to take a newFn if you call it straight away?  Could just take the writer perhaps?

Reply:
Have this ctor to inject mock during tests. The default ctor passes in the public ctor for concrete Writer type.

Reply:
talked offline: moved to take the mock directly, instead of the fn indirection.